runtime/src/iree/schemas/metal_executable_def.fbs

```diff
@@ -12,38 +12,94 @@ namespace iree.hal.metal;
 file_identifier "MTL1";
 file_extension "mtl1";
 
-// A struct for Metal threadgroup size along each dimension.
+// Defines the threadgroup size along each dimension.
 struct ThreadgroupSize {
   x:uint32;
   y:uint32;
   z:uint32;
 }
 
+// Describes the behavior of each buffer binding.
+// Used to populate MTLPipelineBufferDescriptor.
+enum BindingBits:uint64 (bit_flags) {
+  IMMUTABLE = 0,  // 1u << 0
+}
+
+// A compute pipeline (roughly MTLComputePipelineDescriptor).
+table PipelineDef {
+  // Ordinal of the MTLLibrary containing the entry point in the executable
+  // libraries list.
+  library_ordinal:uint32;
+
+  // String name of the entry point MTLFunction in the library.
+  entry_point:string;
+
+  // The maximum number of threads in a threadgroup that can be dispatched to
+  // the compute function. Omit for the default as determined at runtime.
+  // See: maxTotalThreadsPerThreadgroup
+  max_threads_per_threadgroup:uint32;
+
+  // Threadgroup size passed to dispatch commands.
+  threadgroup_size:ThreadgroupSize;
+
+  // Indicates whether the threadgroup size is always a multiple of the thread
+  // execution width.
+  // See: threadGroupSizeIsMultipleOfThreadExecutionWidth
+  threadgroup_size_aligned:bool;
+
+  // Total count in 32-bit words of constants passed to dispatch operations.
+  constant_count:uint32;
+
+  // Buffers used by the pipeline in the order as passed to dispatch operations.
+  binding_flags:[BindingBits];
+
+  // Optional debug information related to the export.
+  debug_info:iree.hal.debug.ExportDef;
+}
+
+// MSL (Metal Shading Language) source code.
+table MSLSourceDef {
+  // MTLLanguageVersion enum indicating the version to interpret the code as.
+  // When omitted the default (latest) version will be used.
+  version:uint32;
+
+  // Source text.
+  code:string;
+
+  // TODO: add compilation options we want to control from the compiler:
+  // https://developer.apple.com/documentation/metal/mtlcompileoptions/4354201-mathmode
+  // https://developer.apple.com/documentation/metal/mtlcompileoptions/4354200-mathfloatingpointfunctions
+  // https://developer.apple.com/documentation/metal/mtlcompileoptions/3564462-preserveinvariance
```

Comment:
Will including these options in the future be a breaking change? If so, can a placeholder/reserved field be added that will make the upgrade path easier?

Reply:
since it's a table they can be appended in a compatible way (the runtime would check for the field presence) - structs in flatbuffers can't be changed without breakage